core/tracing/src/main/scala/org/typelevel/otel4s/trace/SpanBuilder.scala

```diff
@@ -0,0 +1,197 @@
+    * @param spanContext
+    *   the context of the linked span
+    *
+    * @param attributes
+    *   the set of attributes to associate with the link
+    */
+  def withLink(
+      spanContext: SpanContext,
+      attributes: Attribute[_]*
+  ): SpanBuilder[F]
+
+  /** Sets an explicit start timestamp for the newly created span.
+    *
+    * Use this method to specify an explicit start timestamp. If not called, the
+    * implementation will use the timestamp value at ([[createAuto]],
+    * [[createManual]], [[createRes]]) time, which should be the default case.
+    *
+    * '''Note''': the timestamp should be based on `Clock[F].realTime`. Using
+    * `Clock[F].monotonic` may lead to a missing span.
+    *
+    * @param timestamp
+    *   the explicit start timestamp from the epoch
+    */
+  def withStartTimestamp(timestamp: FiniteDuration): SpanBuilder[F]
+
+  /** Sets the finalization strategy for the newly created span.
+    *
+    * The span finalizers are executed upon resource finalization.
+    *
+    * The default strategy is [[SpanFinalizer.Strategy.reportAbnormal]].
+    *
+    * @param strategy
+    *   the strategy to apply upon span finalization
+    */
+  def withFinalizationStrategy(strategy: SpanFinalizer.Strategy): SpanBuilder[F]
+
+  /** Indicates that the span should be the root one and the scope parent should
+    * be ignored.
+    */
+  def root: SpanBuilder[F]
+
+  /** Sets the parent to use from the specified [[SpanContext]]. If not set, the
+    * span that is currently available in the scope will be used as parent.
+    *
+    * '''Note''': if called multiple times, only the last specified value will
+    * be used.
+    *
+    * '''Note''': the previous call of [[root]] will be ignored.
+    *
+    * @param parent
+    *   the span context to use as a parent
+    */
+  def withParent(parent: SpanContext): SpanBuilder[F]
+
+  /** Creates [[Span.Manual]]. The manual span requires to be ended ''explicit''
+    * explicitly.
+    *
+    * The manual span can be used when it's necessary to end a span outside of
+    * the resource scope (i.e. async callback). Make sure the span is ended
+    * properly.
+    *
+    * Leaked span:
+    * {{{
+    * val tracer: Tracer[F] = ???
+    * val leaked: F[Unit] =
+    *   tracer.spanBuilder("manual-span").createManual.use { span =>
+    *     span.setStatus(Status.Ok, "all good")
+    *   }
+    * }}}
+    *
+    * Properly ended span:
+    * {{{
+    * val tracer: Tracer[F] = ???
+    * val ok: F[Unit] =
+    *   tracer.spanBuilder("manual-span").createManual.use { span =>
+    *     span.setStatus(Status.Ok, "all good") >> span.end
+    *   }
+    * }}}
+    *
+    * The finalization strategy is determined by [[SpanFinalizer.Strategy]]. By
+    * default, the abnormal termination (error, cancelation) is recorded.
+    *
+    * @see
+    *   default finalization strategy [[SpanFinalizer.Strategy.reportAbnormal]]
+    */
+  def createManual: Resource[F, Span.Manual[F]]
```

Comment:
Why is this a `Resource` instead of an `F[Span.Manual[F]]`?

Reply:
Does this even need to exist at all?  Is it different from `createAuto.allocated`?  Only thing I can think of is that it gives us a named `end`.

Reply:
`Resource` is needed to apply the finalization strategy during release.

https://github.com/typelevel/otel4s/blob/f2db47d45f7a510eef975fc80470aeab03ddb72f/core/tracing/src/main/scala/org/typelevel/otel4s/trace/SpanBuilder.scala#L92

Also, a manual span is not ended upon resource release. This should be done manually.

The question is the following: do we have a scenario where a span should **outlive** the scope and be ended somewhere after? Nothing comes to my head right away.

We can do the following:
1) Replace `Span.Auto` and `Span.Manual` with `Span`
2) Add the` end` method to `Span`
3) Rename `SpanBuilder#createAuto` to `create`

That way it is possible to end a span earlier.

Reply:
After `end`,

>  Implementations SHOULD ignore all subsequent calls to End and any other Span methods, i.e. the Span becomes non-recording by being ended

For the finalizers to be useful, they need to be sequenced before `end`.  And I don't know why we'd want to call the finalizers, allow other work, and then `end`.  If that's all true, I think all we need is a resource, and "manual" is just an allocated resource.

What gives me pause is that in the spec, `end` can take:

> (Optional) Timestamp to explicitly set the end timestamp. 

I like the idea of hiding `end`, because it changes the state of the span and what's still legal on it, and embedding it in the resource finalizer makes it clear where this happens.  But I'm not sure how to do that and support the explicit end.

What if we had `end` that triggered the finalizer?  "Manual" just returns an `F[Span[F]]`, and automatic returns a `Resource[F, Span[F]]` that invokes `end`?  Then there's still just one kind of Span.

Reply:
Sounds reasonable. There is no sense in invoking finalizers when the span has been terminated. I updated API definitions.